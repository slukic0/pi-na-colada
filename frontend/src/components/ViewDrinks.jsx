import { useEffect, useState, useContext } from 'react';
import ipConstants from '../constants/ipConstants';
import { UserContext } from '../App';
import axios from 'axios';
import {
  CircularProgress,
  Paper,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TableRow,
  IconButton,
} from '@mui/material';
import DeleteIcon from '@mui/icons-material/Delete';

const ViewDrinks = () => {
  const { user } = useContext(UserContext);
  const [isLoading, setIsLoading] = useState(true);
  const [drinks, setDrinks] = useState();

  const getIngredients = async (ids) => {
    const URL = `http://${ipConstants.BACKEND_HOST}/getIngredient`;
    const response = await axios.post(URL, { ingredients: ids });
    return response.data;
  };

  const getDrinks = async (userId) => {
    const URL = `http://${ipConstants.BACKEND_HOST}/getUserDrinks`;
    const response = await axios.post(URL, { id: userId });
    return response.data;
  };

  useEffect(() => {
    const loadData = async () => {
      const drinks = await getDrinks(user.uid);

      // get the ingredient names
      const ingredientIds = [];
      for (const drink of drinks) {
        for (const ingredient of drink.mixture) {
          ingredientIds.push(ingredient.ingredientId);
        }
      }
      const ingredients = await getIngredients(ingredientIds);

      // add names to each drink
      for (const drink of drinks) {
        for (const ingredient of drink.mixture) {
          ingredient.name = ingredients[ingredient.ingredientId].name;
        }
      }
      setDrinks(drinks);
      setIsLoading(false);
    };
    loadData();
  }, []);

  const handleDelete = async (drinkId) => {
    console.log('Delete drink with id ', drinkId);
    const URL = `http://${ipConstants.BACKEND_HOST}/deleteDrink/${drinkId}/${user.uid}`;
    const response = await axios.delete(URL);

    const filtered = drinks.filter((x) => x.id !== drinkId);
    setDrinks(filtered);

    return response.data;
  };

  return isLoading ? (
    <CircularProgress />
  ) : (
    <TableContainer component={Paper}>
      <Table aria-label="simple table">
        <TableHead>
          <TableRow>
            <TableCell align="left">Name</TableCell>
            <TableCell align="right">Mixture</TableCell>
            <TableCell align="center">Actions</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {drinks.map((drink) => (
            <TableRow key={drink.id} sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
              <TableCell component="th" scope="row">
                {drink.name}
              </TableCell>
              <TableCell align="right">{drink.mixture.map((x) => x.name).join(', ')}</TableCell>
              <TableCell align="center">
                <IconButton onClick={() => handleDelete(drink.id)}>
                  <DeleteIcon />
                </IconButton>
              </TableCell>
            </TableRow>
          ))}
        </TableBody>
      </Table>
    </TableContainer>
  );
};

export default ViewDrinks;
