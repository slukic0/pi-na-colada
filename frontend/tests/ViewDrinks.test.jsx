import { render, screen, waitFor } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import axios from 'axios';
import { act } from 'react-dom/test-utils';
import ViewDrinks from '../src/components/ViewDrinks';
import UserContextWrapper from './UserContextWrapper';

//import { mock } from '@testing-library/react';

vi.mock('axios');

describe('view ingredient', () => {
  beforeEach(() => {
    axios.post.mockImplementation((url) => {
      if (url.includes('getUserDrinks')) {
        return Promise.resolve({
          data: [
            {
              id: 'testDrinkId',
              mixture: [
                {
                  amount: 1,
                  ingredientId: 'ingId1',
                },
                {
                  amount: 2,
                  ingredientId: 'ingId2',
                },
              ],
              name: 'Test Drink',
            },
          ],
        });
      } else if (url.includes('getIngredient')) {
        return Promise.resolve({
          data: {
            ingId1: { name: 'ingredient1 name' },
            ingId2: { name: 'ingredient2 name' },
          },
        });
      }
    });
  });
  it('should render the table of ingredients', async () => {
    act(() => {
      render(<UserContextWrapper child={<ViewDrinks />} user={{ id: 'userId' }} />);
    });

    await waitFor(() => {
      expect(screen.getByText('Test Drink')).toBeInTheDocument();
      expect(screen.getByText('ingredient1 name, ingredient2 name')).toBeInTheDocument();
    });
  });

  it('should match its snapshot', async () => {
    let tree;
    await act(async () => {
      tree = render(<UserContextWrapper child={<ViewDrinks />} user={{ id: 'userId' }} />);
    });
    await waitFor(() => {
      expect(tree).toMatchSnapshot();
    });
  });
});
