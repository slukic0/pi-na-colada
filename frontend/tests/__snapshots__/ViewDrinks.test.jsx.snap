// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`view ingredient > should match its snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      >
        <table
          aria-label="simple table"
          class="MuiTable-root css-rqglhn-MuiTable-root"
        >
          <thead
            class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
          >
            <tr
              class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
                scope="col"
              >
                Name
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
                scope="col"
              >
                Mixture
              </th>
              <th
                class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
                scope="col"
              >
                Actions
              </th>
            </tr>
          </thead>
          <tbody
            class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
          >
            <tr
              class="MuiTableRow-root css-34nofg-MuiTableRow-root"
            >
              <th
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
                scope="row"
              >
                Test Drink
              </th>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
              >
                ingredient1 name, ingredient2 name
              </td>
              <td
                class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
              >
                <button
                  class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                  tabindex="0"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="DeleteIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                    />
                  </svg>
                  <span
                    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                  />
                </button>
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
    >
      <table
        aria-label="simple table"
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
              scope="col"
            >
              Mixture
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1ndpvdd-MuiTableCell-root"
              scope="col"
            >
              Actions
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root css-34nofg-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
              scope="row"
            >
              Test Drink
            </th>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            >
              ingredient1 name, ingredient2 name
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignCenter MuiTableCell-sizeMedium css-1yhpg23-MuiTableCell-root"
            >
              <button
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                tabindex="0"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="DeleteIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                  />
                </svg>
                <span
                  class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                />
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
